### PR TITLE
S28-2252 share a booking

### PR DESCRIPTION
**Pull request by LaurenceRawlings**
### JIRA link ###

https://tools.hmcts.net/jira/browse/S28-2252

### Change description ###

- Created endpoint to share a booking

```
/bookings/{bookingid}/share
```

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/preapi/services/RecordingService.java [hunk 1]
```diff
@@ -124,9 +136,44 @@ public void deleteById(UUID bookingId, UUID recordingId) {
...
+
+        if (captureSession == null) {
+            throw new NotFoundException("Capture Session: " + shareRecordingDTO.getCaptureSessionId());
+        }
+        if (sharedByUser == null) {
+            throw new NotFoundException("Shared by User: " + shareRecordingDTO.getSharedByUserId());
+        }
+        if (sharedWithUser == null) {
+            throw new NotFoundException("Shared with User: " + shareRecordingDTO.getSharedWithUserId());
+        }
```
Comment (lucas-phillips28) [inline src/main/java/uk/gov/hmcts/reform/preapi/services/RecordingService.java:156]:
```suggestion
        var captureSession = captureSessionRepository.findById(shareRecordingDTO.getCaptureSessionId()).orElseThrow(() -> new NotFoundException("Capture Session: " + shareRecordingDTO.getCaptureSessionId()));
        var sharedByUser = userRepository.findById(shareRecordingDTO.getSharedByUserId()).orElseThrow(() -> new NotFoundException("Shared by User: " + shareRecordingDTO.getSharedByUserId()));
        var sharedWithUser = userRepository.findById(shareRecordingDTO.getSharedWithUserId()).orElseThrow(() -> new NotFoundException("Shared with User: " + shareRecordingDTO.getSharedWithUserId()));
```

File: src/main/java/uk/gov/hmcts/reform/preapi/controllers/RecordingController.java [hunk 2]
```diff
@@ -78,4 +86,18 @@ public ResponseEntity<Void> deleteRecordingById(
         recordingService.deleteById(bookingId, recordingId);
         return ResponseEntity.ok().build();
     }
+
+    @PutMapping("/{recordingId}/share")
```
Comment (lucas-phillips28) [inline src/main/java/uk/gov/hmcts/reform/preapi/controllers/RecordingController.java:90]:
```suggestion
    @PutMapping("/{recordingId}/share")
    @Operation(operationId = "shareRecordingById", summary = "Share a Recording")
```

File: src/main/java/uk/gov/hmcts/reform/preapi/controllers/RecordingController.java [hunk 3]
```diff
@@ -84,4 +85,19 @@ public ResponseEntity<Void> deleteRecordingById(
...
+    @Operation(operationId = "shareRecordingById", summary = "Share a Recording")
+    public ResponseEntity<Void> shareRecordingById(
+        @PathVariable UUID bookingId,
+        @PathVariable UUID recordingId,
+        @RequestBody ShareRecordingDTO shareRecordingDTO
+    ) {
+        // TODO Ensure user has access to share the recording
+        if (!recordingId.equals(shareRecordingDTO.getCaptureSessionId())) {
```
Comment (lucas-phillips28) [inline src/main/java/uk/gov/hmcts/reform/preapi/controllers/RecordingController.java:97]: Shouldn't the recording id relate to a recording. Then you could get the capture session related to the recording.

Comment (lucas-phillips28) [reply src/main/java/uk/gov/hmcts/reform/preapi/controllers/RecordingController.java:97]: So you would do something like:
1.  Get the recording by the recording id
2. Get the related capture session's id for the recording
3. Check the id found matches the id in the dto.

File: src/main/java/uk/gov/hmcts/reform/preapi/services/RecordingService.java [hunk 4]
```diff
@@ -26,13 +29,19 @@ public class RecordingService {
     private final RecordingRepository recordingRepository;
     private final BookingRepository bookingRepository;
     private final CaptureSessionRepository captureSessionRepository;
+    private final ShareRecordingRepository shareRecordingRepository;
```
Comment (lucas-phillips28) [inline src/main/java/uk/gov/hmcts/reform/preapi/services/RecordingService.java:32]: You are going to need some mock beans in the tests for the service (file called `RecordingServiceTest`) 
```java
    @MockBean
    private ShareRecordingRepository shareRecordingRepository;

    @MockBean
    private UserRepository userRepository;
```

File: src/main/java/uk/gov/hmcts/reform/preapi/repositories/ShareRecordingRepository.java [hunk 5]
```diff
@@ -9,4 +9,5 @@
 
 @Repository
 public interface ShareRecordingRepository extends JpaRepository<ShareRecording, UUID> {
+    boolean existsByIdAndDeletedAtIsNull(UUID id);
```
Comment (lucas-phillips28) [inline src/main/java/uk/gov/hmcts/reform/preapi/repositories/ShareRecordingRepository.java:12]: I don't think this function is being used. Is this going to be used or can it be deleted?

File: src/main/java/uk/gov/hmcts/reform/preapi/controllers/RecordingController.java (hunk 3 above)
Comment (LaurenceRawlings) [reply src/main/java/uk/gov/hmcts/reform/preapi/controllers/RecordingController.java]: Sharing a booking now instead

File: src/test/java/uk/gov/hmcts/reform/preapi/controller/BookingControllerTest.java [hunk 6]
```diff
@@ -358,6 +360,83 @@ void deleteBookingEndpoint202() throws Exception {
...
+        shareBooking.setBookingId(bookingId);
+
+        var mockCaseDTO = new CaseDTO();
+        mockCaseDTO.setId(caseId);
+
+        var mockBookingDTO = new BookingDTO();
+        mockBookingDTO.setId(bookingId);
+
+        when(caseService.findById(caseId)).thenReturn(mockCaseDTO);
+        when(bookingService.findById(bookingId)).thenReturn(mockBookingDTO);
```
Comment (lucas-phillips28) [inline src/test/java/uk/gov/hmcts/reform/preapi/controller/BookingControllerTest.java:384]:
```suggestion
        when(bookingService.shareBookingById(any())).thenReturn(UpsertResult.CREATED);
```

File: src/test/java/uk/gov/hmcts/reform/preapi/controller/BookingControllerTest.java [hunk 7]
```diff
@@ -358,6 +360,83 @@ void deleteBookingEndpoint202() throws Exception {
...
+        shareRecording.setSharedByUserId(sharedByUserId);
+
+        var mockCaseDTO = new CaseDTO();
+        mockCaseDTO.setId(caseId);
+
+        var mockBookingDTO = new BookingDTO();
+        mockBookingDTO.setId(bookingId);
+
+        when(caseService.findById(caseId)).thenReturn(mockCaseDTO);
+        when(bookingService.findById(bookingId)).thenReturn(mockBookingDTO);
```
Comment (lucas-phillips28) [inline src/test/java/uk/gov/hmcts/reform/preapi/controller/BookingControllerTest.java:424]: These aren't needed

File: src/main/java/uk/gov/hmcts/reform/preapi/repositories/ShareRecordingRepository.java (hunk 5 above)
Comment (LaurenceRawlings) [reply src/main/java/uk/gov/hmcts/reform/preapi/repositories/ShareRecordingRepository.java]: Removed

File: src/main/java/uk/gov/hmcts/reform/preapi/services/BookingService.java [hunk 8]
```diff
@@ -97,4 +107,23 @@ public void markAsDeleted(UUID id) {
...
+            .orElseThrow(() -> new NotFoundException("Shared with User: " + shareBookingDTO.getSharedWithUserId()));
+
+        var shareBookingEntity = new ShareBooking();
+        shareBookingEntity.setId(shareBookingDTO.getId());
+        shareBookingEntity.setBooking(booking);
+        shareBookingEntity.setSharedBy(sharedByUser);
+        shareBookingEntity.setSharedWith(sharedWithUser);
+        shareBookingRepository.save(shareBookingEntity);
+
+        return UpsertResult.CREATED;
```
Comment (lucas-phillips28) [inline src/main/java/uk/gov/hmcts/reform/preapi/services/BookingService.java:127]: Are we only allowing create on this endpoint? It makes sense but if we are we probably want an extra check to make sure we aren't accidentally updating.
```java
if(shareBookingRepository.existsById(shareBookingDTO.getId()) {
    // throw an error
}
```

File: src/main/java/uk/gov/hmcts/reform/preapi/services/BookingService.java [hunk 9]
```diff
@@ -114,4 +125,27 @@ public void markAsDeleted(UUID id) {
...
+    @Transactional
+    public UpsertResult shareBookingById(ShareBookingDTO shareBookingDTO) {
+        final var booking = bookingRepository.findById(shareBookingDTO.getBookingId())
+            .orElseThrow(() -> new NotFoundException("Booking: " + shareBookingDTO.getBookingId()));
+        final var sharedByUser = userRepository.findById(shareBookingDTO.getSharedByUserId())
+            .orElseThrow(() -> new NotFoundException("Shared by User: " + shareBookingDTO.getSharedByUserId()));
+        final var sharedWithUser = userRepository.findById(shareBookingDTO.getSharedWithUserId())
+            .orElseThrow(() -> new NotFoundException("Shared with User: " + shareBookingDTO.getSharedWithUserId()));
+
+        if (shareBookingRepository.existsById(shareBookingDTO.getId())) {
```
Comment (lucas-phillips28) [inline src/main/java/uk/gov/hmcts/reform/preapi/services/BookingService.java:138]: Could move this to the top of the function

**Review by lucas-phillips28 — DISMISSED** [review]
Looks good to me. One small comment but not completely necessary

**Review by jasonpaige — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/preapi/services/BookingService.java (hunk 9 above)
Comment (LaurenceRawlings) [reply src/main/java/uk/gov/hmcts/reform/preapi/services/BookingService.java]: Done

**Review by lucas-phillips28 — APPROVED** [review]

**Review by jasonpaige — APPROVED** [review]

**Pull request merged**